src/core/ddsi/src/ddsi_serdata_cdr.c

```diff
@@ -0,0 +1,361 @@
+/*
+ * Copyright(c) 2006 to 2022 ZettaScale Technology and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <stdlib.h>
+#include <ctype.h>
+#include <assert.h>
+#include <string.h>
+
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/log.h"
+#include "dds/ddsrt/md5.h"
+#include "dds/ddsrt/mh3.h"
+#include "dds/ddsi/q_bswap.h"
+#include "dds/ddsi/ddsi_config_impl.h"
+#include "dds/ddsi/q_freelist.h"
+#include "dds/ddsi/ddsi_tkmap.h"
+#include "dds/cdr/dds_cdrstream.h"
+#include "dds/ddsi/q_radmin.h"
+#include "dds/ddsi/ddsi_domaingv.h"
+#include "dds/ddsi/ddsi_serdata_cdr.h"
+
+#define DEFAULT_NEW_SIZE 128
+#define CHUNK_SIZE 128
+
+#ifndef NDEBUG
+static int ispowerof2_size (size_t x)
+{
+  return x > 0 && !(x & (x-1));
+}
+#endif
+
+static size_t alignup_size (size_t x, size_t a)
+{
+  size_t m = a - 1;
+  assert (ispowerof2_size (a));
+  return (x + m) & ~m;
+}
+
+static void *serdata_cdr_append (struct ddsi_serdata_cdr **d, size_t n)
+{
+  char *p;
+  if ((*d)->pos + n > (*d)->size)
+  {
+    size_t size1 = alignup_size ((*d)->pos + n, CHUNK_SIZE);
+    *d = ddsrt_realloc (*d, offsetof (struct ddsi_serdata_cdr, data) + size1);
+    (*d)->size = (uint32_t)size1;
+  }
+  assert ((*d)->pos + n <= (*d)->size);
+  p = (*d)->data + (*d)->pos;
+  (*d)->pos += (uint32_t)n;
+  return p;
+}
+
+static void serdata_cdr_append_blob (struct ddsi_serdata_cdr **d, size_t sz, const void *data)
+{
+  char *p = serdata_cdr_append (d, sz);
+  memcpy (p, data, sz);
+}
+
+static struct ddsi_serdata *fix_serdata_cdr (struct ddsi_serdata_cdr *d, uint32_t basehash)
+{
+  d->c.hash = basehash;
+  return &d->c;
+}
+
+static uint32_t serdata_cdr_get_size (const struct ddsi_serdata *dcmn)
+{
+  const struct ddsi_serdata_cdr *d = (const struct ddsi_serdata_cdr *) dcmn;
+  return d->pos + (uint32_t)sizeof (struct dds_cdr_header);
+}
+
+static bool serdata_cdr_eqkey (const struct ddsi_serdata *acmn, const struct ddsi_serdata *bcmn)
+{
+  (void) acmn;
```

Comment:
I think it would make sense to add a comment somewhere that this is specialised for types that have no key fields.

Reply:
I've added a comment on the struct definition for `sertype_cdr` and a comment in this function ([fab4340](https://github.com/eclipse-cyclonedds/cyclonedds/pull/1468/commits/fab4340750acdb33031ae156160d16b3cb67c7b5))